DietPi-Software | rTorrent

Joulinar opened the pull request:
Switch rTorrent/Lighttpd to a new/own configuration file

**Status**: Ready

- [x] adjust `dietpi-software` to allow new rTorrent configuration file for `Lighttpd`
- [x] check new install
- [x] check uninstall new config
- [x] check reinstall and conversion from old to new configuration
- [x] check uninstall using old configuration

**Reference**: https://github.com/MichaIng/DietPi/issues/4330

**Commit list/description**:
- DietPi-Software | rTorrent - rework configuration for Lighttpd and enable own configuration file
- DietPi-Software | rTorrent - ensure on reinstall old configuration to be deleted and new configuration file to be used
- DietPi-Software | rTorrent - on uninstall respect old and new configuration file
- DietPi-Software | rTorrent - Fix failing reinstall by replacing wildcard asterisk with actual directory name, derived from version string, like we do in case of Box86: Actual version string is tag name minus leading "v".

Joulinar committed:
v7.2

DietPi-Software | rTorrent - rework configuration for Lighttpd and enable own configuration file
DietPi-Software | rTorrent - ensure on reinstall old configuration to be deleted and new configuration file to be used
DietPi-Software | rTorrent - on uninstall respect old and new configuration file

Joulinar commented:
@MichaIng 
there is an issue on reinstall. But this seems to be there all the time already

```
[  OK  ] DietPi-Software | tar xf v3.10.tar.gz
[  OK  ] DietPi-Software | rm v3.10.tar.gz
[FAILED] DietPi-Software | mv ruTorrent-3.10/conf/config.php ruTorrent-*/conf/config.php.example
[ INFO ] DietPi-Software | Executing alternative command: mv ruTorrent-3.10/conf/config.php /var/www/rutorrent/conf/config.php.example
[  OK  ] Alternative command execution | Completed
[FAILED] DietPi-Software | mv ruTorrent-3.10/conf/access.ini ruTorrent-*/conf/access.ini.example
[ INFO ] DietPi-Software | Executing alternative command: mv ruTorrent-3.10/conf/access.ini /var/www/rutorrent/conf/access.ini.example
[  OK  ] Alternative command execution | Completed
[FAILED] DietPi-Software | mv ruTorrent-3.10/conf/plugins.ini ruTorrent-*/conf/plugins.ini.example
[ INFO ] DietPi-Software | Executing alternative command: mv ruTorrent-3.10/conf/plugins.ini /var/www/rutorrent/conf/plugins.ini.example
```

It's even failing using the original script
`mv ruTorrent-3.10/conf/config.php ruTorrent-*/conf/config.php.example`

but it should be 
`mv ruTorrent-3.10/conf/config.php /var/www/rutorrent/conf/config.php.example`
isn't it?

This is the code block in question
https://github.com/MichaIng/DietPi/blob/rtorrent/dietpi/dietpi-software#L5907-L5910

MichaIng commented:
Hmm, the code is as intended, first the files should be moved aside in the RAM dir, afterwards the whole RAM dir is moved to disk. But the asterisk `*` glob does not expand when the target file (or dir) does not exist.

We can replace the asterisk with `${version#v}` (above obtained version string with leading `v` removed).

Joulinar commented:
Not sure if we misunderstood but issue is with the target directory. It should be `/var/www/rutorrent/` and not `ruTorrent-*/`

MichaIng commented:
The target directory was meant to be the same as the origin, so the file should just be renamed in place to not overwrite the installed one, when all is moved to disk, but serve as updated example. The `*` is the globbing wild card, so it matches all existing files and dirs. That works for the source file, as it exist, but not for the target file, as it does not exist (even that the path that shall be expanded does).
```
mv ruTorrent-3.10/conf/config.php ruTorrent-*/conf/config.php.example
```
- The `*` becomes `3.10` for the source, but stays `*` for the target, as the whole final file path needs to exist for any glob expansion.

The syntax is a bit un-intuitive at this point, hence the bug.

My idea was now to use the actual version string to not rely on globbing. Your solution would change the order of the intended steps, but the final result will be the same indeed. But avoiding globbing in such cases is good anyway, as it can theoretically match two files/dirs, which then would cause havok, and I'm fan of being explicit, where possible.

Joulinar commented:
I need to say you lost me. Do we keep it as is or do we need to change thinks ?

MichaIng commented:
Sorry 😄, I'll apply the change how I mean it.

MichaIng committed:
v7.2

+ DietPi-Software | rTorrent: Fix failing reinstall by replacing wildcard asterisk with actual directory name, derived from version string, like we do in case of Box86: Actual version string is tag name minus leading "v".

MichaIng commented:
Done: https://github.com/MichaIng/DietPi/pull/4332/commits/5fa44582647163c34343e2a6ea4d1bec970d70ca
To verify whether this has some systematic, I found we do the same in case of Box86 (`${version#v}`), so it seems to be some systematic that it's removed from the tag name for the directory name in the archive (`ruTorrent-3.10`, not `ruTorrent-v3.10`).

Joulinar commented:
AHH now I got the logic. You are going to move config files inside `/tmp` first and copy it afterwards to target location. OK I misunderstood this completely. Thought it should be copied directly. 🤣 

Your change is working now

```
[  OK  ] DietPi-Software | tar xf v3.10.tar.gz
[  OK  ] DietPi-Software | rm v3.10.tar.gz
[  OK  ] DietPi-Software | mv ruTorrent-3.10/conf/config.php ruTorrent-3.10/conf/config.php.example
[  OK  ] DietPi-Software | mv ruTorrent-3.10/conf/access.ini ruTorrent-3.10/conf/access.ini.example
[  OK  ] DietPi-Software | mv ruTorrent-3.10/conf/plugins.ini ruTorrent-3.10/conf/plugins.ini.example
[  OK  ] DietPi-Software | cp -a ruTorrent-3.10/. /var/www/rutorrent/
[  OK  ] DietPi-Software | rm -R ruTorrent-3.10
```

Ready from my side for your final quality check.

MichaIng committed:
v7.2

+ DietPi-Software | rTorrent: Do not attempt do enable/disable Lighttpd configs when they are enabled/disabled already, sine lighty-enable-mod/lighty-disable-mod is picky in those cases and returns an error code. Also generally allow to proceed an uninstall even if en error happens.

MichaIng commented:
I added some checks when running lighty-enable-mod/lighty-disable-mod, as those will error out when the config is enabled/disabled already. I'll do another install + reinstall test.

MichaIng commented:
Okay, fresh install, upgrade from prior to the change, reinstall and uninstall work all fine. I'll merge.
